src/twisted/internet/cfreactor.py

```diff
@@ -167,36 +167,41 @@ def _socketCallback(self, cfSocket, callbackType,
             )
             return
 
-        why = None
-        isRead = False
-        src, skt, readWriteDescriptor, rw = self._fdmap[fd]
-        try:
-            if readWriteDescriptor.fileno() == -1:
-                why = _NO_FILEDESC
-            else:
-                isRead = callbackType == kCFSocketReadCallBack
-                # CFSocket seems to deliver duplicate read/write notifications
-                # sometimes, especially a duplicate writability notification
-                # when first registering the socket.  This bears further
-                # investigation, since I may have been mis-interpreting the
-                # behavior I was seeing. (Running the full Twisted test suite,
-                # while thorough, is not always entirely clear.) Until this has
-                # been more thoroughly investigated , we consult our own
-                # reading/writing state flags to determine whether we should
-                # actually attempt a doRead/doWrite first.  -glyph
-                if isRead:
-                    if rw[_READ]:
-                        why = log.callWithLogger(
-                            readWriteDescriptor, readWriteDescriptor.doRead)
+        def _drdw():
+
+            why = None
+            isRead = False
+            src, skt, readWriteDescriptor, rw = self._fdmap[fd]
```

Comment:
I don't like the fact that `readWriteDescriptor` is bound twice, once inside and once outside.  It's never _re_-bound, so could you just move this tuple-unpacking assignment outside of `_drdw`?